File: aurora-scala/src/main/scala/net/gree/aurora/scala/domain/datasource/DataSourceSelector.scala
package net.gree.aurora.scala.domain.datasource

import net.gree.aurora.domain.datasource.{DataSource => JDataSource}
import net.gree.aurora.domain.datasource.{DataSourceSelector => JDataSourceSelector, DataSourceSelectorFactory}
import scala.util.Try

/**
 * [[net.gree.aurora.domain.datasource.DataSource]]のScala版。
 *
 * @tparam T ヒントの型
 */
trait DataSourceSelector[T] {

  /**
   * ヒントから[[net.gree.aurora.scala.domain.datasource.DataSource]]を選択する。
   *
   * @param hint ヒント
   * @return Tryでラップされた[[net.gree.aurora.scala.domain.datasource.DataSource]]
   */
  def selectDataSource(hint: Option[T] = None): Try[DataSource]

  /**
   * ヒントから[[net.gree.aurora.scala.domain.datasource.JDBCDataSource]]を選択する。
   *
   * @param hint ヒント
   * @return Tryでラップされた[[net.gree.aurora.scala.domain.datasource.JDBCDataSource]]
   */
  def selectDataSourceAsJDBC(hint: Option[T] = None): Try[JDBCDataSource]

  /**
   * ヒントから[[net.gree.aurora.scala.domain.datasource.RedisDataSource]]を選択する。
   *
   * @param hint ヒント
   * @return Tryでラップされた[[net.gree.aurora.scala.domain.datasource.RedisDataSource]]
   */
  def selectDataSourceAsRedis(hint: Option[T] = None): Try[RedisDataSource]

}

/**
 * コンパニオンオブジェクト。
 */
object DataSourceSelector {

  /**
   * JavaオブジェクトからScalaオブジェクトを生成する。
   *
   * @param underlying [[net.gree.aurora.domain.datasource.DataSourceSelector]]
   * @tparam T ヒントの型
   * @return [[net.gree.aurora.scala.domain.datasource.DataSourceSelector]]
   */
  private[scala] def apply[T](underlying: JDataSourceSelector[T]): DataSourceSelector[T] =
    new DataSourceSelectorImpl[T](underlying)

  /**
   * ファクトリメソッド。
   *
   * @param dataSourceRepository [[net.gree.aurora.scala.domain.datasource.DataSourceRepository]]
   * @param dataSourceIdResolver [[net.gree.aurora.scala.domain.datasource.DataSourceIdResolver]]
   * @tparam T ヒントの型
   * @return [[net.gree.aurora.scala.domain.datasource.DataSourceSelector]]
   */
  def apply[T](dataSourceRepository: DataSourceRepository, dataSourceIdResolver: DataSourceIdResolver[T]): DataSourceSelector[T] =
    apply(DataSourceSelectorFactory.create(dataSourceRepository, dataSourceIdResolver))

}

File: aurora-core/src/main/java/net/gree/aurora/domain/datasource/DataSourceSelector.java
package net.gree.aurora.domain.datasource;

import org.sisioh.dddbase.utils.Try;

/**
 * 複数の{@link DataSource}から一つの{@link DataSource}を選択するためのサービス。
 *
 * @param <T> ヒントの型
 */
public interface DataSourceSelector<T> {

    /**
     * ヒントを元に{@link DataSource}の派生型を選択する。
     *
     * @param clazz {@link DataSource}の派生クラス
     * @param hint  ヒント
     * @param <A>   派生型
     * @return Tryでラップされた{@link DataSource}
     */
    <A extends DataSource> Try<A> selectDataSourceAs(Class<A> clazz, T hint);

    /**
     * ヒントを元に{@link DataSource}を選択する。
     *
     * @param hint ヒント
     * @return Tryでラップされた{@link DataSource}
     */
    Try<DataSource> selectDataSource(T hint);

    /**
     * ヒントを元に{@link JDBCDataSource}を選択する。
     *
     * @param hint ヒント
     * @return Tryでラップされた{@link JDBCDataSource}
     */
    Try<JDBCDataSource> selectDataSourceAsJDBC(T hint);

    /**
     * ヒントを元に{@link RedisDataSource}を選択する。
     *
     * @param hint ヒント
     * @return Tryでラップされた{@link RedisDataSource}
     */
    Try<RedisDataSource> selectDataSourceAsRedis(T hint);

}
